core/templates/pages/exploration-editor-page/services/changes-in-human-readable-form.service.spec.ts

```diff
@@ -0,0 +1,632 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for ChangesInHumanReadableFormService.
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { ChangesInHumanReadableFormService } from
+// eslint-disable-next-line max-len
+  'pages/exploration-editor-page/services/changes-in-human-readable-form.service';
+import { OutcomeObjectFactory } from
+  'domain/exploration/OutcomeObjectFactory';
+
+describe('Changes In Human Readable Form Service', () => {
+  let cihrfs, oof;
+
+  beforeEach(() => {
+    cihrfs = TestBed.get(ChangesInHumanReadableFormService);
+    oof = TestBed.get(OutcomeObjectFactory);
+  });
+
+  it('should make rules list human readable', () => {
+    const answerGroupValue = {
+      rules: [{
+        type: 'Type1',
+        inputs: {
+          input1: 'input1',
+          input2: 'input2'
+        }
+      }]
+    };
+
+    cihrfs.makeRulesListHumanReadable(answerGroupValue)
+      .forEach((element: HTMLElement, i) => {
+        const rule = answerGroupValue.rules[i];
+        expect(element.outerHTML).toBe(
+          '<li>' +
+          '<p>Type: ' + rule.type + '</p>' +
+          '<p>Value: ' + Object.keys(rule.inputs).toString() + '</p>' +
+          '</li>'
+        );
+      });
+  });
+
+  it('should not make rules list human readable when there are no rules',
+    () => {
+      const answerGroupValue = {
+        rules: []
+      };
+      expect(cihrfs.makeRulesListHumanReadable(answerGroupValue)).toEqual([]);
+    });
+
+  it('should get state property value when state is an array', () => {
+    const state = ['value1', 'value2'];
+    expect(cihrfs.getStatePropertyValue(state)).toBe('value2');
+  });
+
+  it('should get state property value when state is an object', () => {
```

Comment:
This test and the above one I don't understand... Wouldn't state usually be an object itself?

Reply:
state can be either an array or an object. These two tests get the property value (described in the service function) accordingly to its type and evaluate it.